File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

while 1:
    s = sys.stdin.readline().rstrip()
    if s[0] == '.' and len(s) == 1:
        break
    stack = []

    sign = True

    for a in s:
        if a == '(':
            stack.append(a)
        elif a == '[':
            stack.append(a)
        elif a == ')':
            if len(stack) > 0:
                if stack[-1] == "(":
                    stack.pop()
                else:
                    sign = False
                    break
            else:
                sign = False
                break
        elif a == ']':
            if len(stack) > 0:
                if stack[-1] == "[":
                    stack.pop()
                else:
                    sign = False
                    break
            else:
                sign = False
                break

    if sign and len(stack) == 0:
        print('yes')
    else:
        print('no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String s;

        while (!".".equals((s = br.readLine()))) {
            Stack<Character> stack = new Stack<>();
            boolean sign = true;

            for (char c : s.toCharArray()) {
                if (c == '(') {
                    stack.push(c);
                } else if (c == '[') {
                    stack.push(c);
                } else if (c == ')') {
                    if (stack.isEmpty() || stack.peek() != '(') {
                        sign = false;
                        break;
                    } else {
                        stack.pop();
                    }
                } else if (c == ']') {
                    if (stack.isEmpty() || stack.peek() != '[') {
                        sign = false;
                        break;
                    } else {
                        stack.pop();
                    }
                }
            }
            if (stack.isEmpty() && sign) {
                sb.append("yes").append('\n');
            } else {
                sb.append("no").append('\n');
            }
        }
        System.out.println(sb);



    }

}

